repair/repair.cc

```diff
@@ -1262,6 +1262,60 @@ struct repair_options {
         boost::split(var, it->second, boost::algorithm::is_any_of(","));
         options.erase(it);
     }
+
+public:
+    // Returns a list of hosts participating in repair
+    std::vector<gms::inet_address> get_participating_hosts(const database& db) const {
```

Comment:
The implementation looks incorrect to me. We can not just look at the hosts and data_centers to figure out the repair followers. In most cases the hosts and data_centers will be empty. Repair calculate the participants on a vnode range basis. Take a look at: get_neighbors.

Reply:
Another thing to note: when a repair is started by node operations, it uses sync_data_using_repair(). The caller can provides the neighbors, take a look at:
```
repair_neighbors repair_info::get_repair_neighbors(const dht::token_range& range) {
    return neighbors.empty() ?
        repair_neighbors(get_neighbors(db.local(), keyspace, range, data_centers, hosts, ignore_nodes)) :
        neighbors[range];
}
```

Reply:
> The implementation looks incorrect to me. We can not just look at the hosts and data_centers to figure out the repair followers. In most cases the hosts and data_centers will be empty. Repair calculate the participants on a vnode range basis. Take a look at: get_neighbors.

I was, in fact, looking at the `get_neighbors` function when I was writing `get_participating_hosts` :)

When I was reading the `get_neighbors` function, I inferred that it does the following:

1. Calculate natural endpoints for `tok`,
1. Filter out endpoints based on `hosts`, `data_centers` and `ignore_nodes` arguments,
1. Return remaining endpoints.

To me, it looks like the second step just filters out nodes that aren't participating in repair. Based on that, I figured out that the `get_participating_hosts` function could work like this:

1. Get the list of all nodes in the cluster,
1. Filter out endpoints (works just like the step 2. of `get_neighbors`),
1. Return remaining endpoints.

The final version of `get_participating_hosts` works a bit differently (it doesn't always have to construct a list of all nodes and then filter them out), but its result should be equivalent to the algorithm above.

Reply:
Why do you do "Get the list of all nodes in the cluster"? Don't we only want to ask the nodes that participating the repair to receive the hints?

Say we have n1,n2,n3,n4,n5 in the cluster. Now user starts a repair for a given table and ranges. Only n1,n2,n3 are participating the repair.  Don't we want only n1, n2, n3 to receive the hints from the cluster.

Since, it is possible that all the 5 nodes may contain hints targeted to n1, n2, n3. We need to ask all the 5 nodes to flush the hints targeted to n1, n2, n3.

Is this what we want?

The current code storage_proxy::wait_for_hints_to_be_replayed and get_participating_hosts do not do this.

Reply:
> Why do you do "Get the list of all nodes in the cluster"? Don't we only want to ask the nodes that participating the repair to receive the hints?
> 
> Say we have n1,n2,n3,n4,n5 in the cluster. Now user starts a repair for a given table and ranges. Only n1,n2,n3 are participating the repair. Don't we want only n1, n2, n3 to receive the hints from the cluster.
> 
> Since, it is possible that all the 5 nodes may contain hints targeted to n1, n2, n3. We need to ask all the 5 nodes to flush the hints targeted to n1, n2, n3.
> 
> Is this what we want?
> 
> The current code storage_proxy::wait_for_hints_to_be_replayed and get_participating_hosts do not do this.

You are right. I updated the implementation so that it asks _all_ nodes in the cluster to replay their hints towards nodes participating in repair.

Reply:
I looked the latest version. I think it is still incorrect. get_participating_hosts returns all nodes in the cluster. This is not what we want. We need to get the real repair followers.

Reply:
we can do this:

```
unordered_set<> participants
for all the ranges to repair
      nodes = get_neighbors(range)
      add nodes to participants set 
```

get_participating_hosts is duplicating  existing get_neighbors.